### ISSUE TITLE
0.20.3 fails to work in python 2.7

### ISSUE THREAD

**Issue by MatzeB** [dulwich/dulwich#783]
I get this error when trying to install dulwish 0.20.3 with python 2.7:
```
[matthiasb@matthiasb-mbp] /opt/facebook > sudo /opt/homebrew/bin/pip2.7 install dulwich
Downloading/unpacking dulwich
  Downloading dulwich-0.20.3.tar.gz (376kB): 376kB downloaded
  Running setup.py (path:/private/tmp/pip_build_root/dulwich/setup.py) egg_info for package dulwich
    /opt/homebrew/Cellar/python27/2.7.10_2/Frameworks/Python.framework/Versions/2.7/lib/python2.7/distutils/dist.py:267: UserWarning: Unknown distribution option: 'project_urls'
      warnings.warn(msg)
    /opt/homebrew/Cellar/python27/2.7.10_2/Frameworks/Python.framework/Versions/2.7/lib/python2.7/distutils/dist.py:267: UserWarning: Unknown distribution option: 'python_requires'
      warnings.warn(msg)

    warning: no files found matching 'appveyor.yml'
    warning: no files found matching '.travis.yml'
Downloading/unpacking urllib3>=1.24.1 (from dulwich)
  Downloading urllib3-1.25.9-py2.py3-none-any.whl (126kB): 126kB downloaded
Downloading/unpacking certifi (from dulwich)
  Downloading certifi-2020.4.5.2-py2.py3-none-any.whl (157kB): 157kB downloaded
Installing collected packages: dulwich, urllib3, certifi
  Running setup.py install for dulwich
    /opt/homebrew/Cellar/python27/2.7.10_2/Frameworks/Python.framework/Versions/2.7/lib/python2.7/distutils/dist.py:267: UserWarning: Unknown distribution option: 'project_urls'
      warnings.warn(msg)
    /opt/homebrew/Cellar/python27/2.7.10_2/Frameworks/Python.framework/Versions/2.7/lib/python2.7/distutils/dist.py:267: UserWarning: Unknown distribution option: 'python_requires'
      warnings.warn(msg)

    warning: no files found matching 'appveyor.yml'
    warning: no files found matching '.travis.yml'
    building 'dulwich._objects' extension
    clang -fno-strict-aliasing -fno-common -dynamic -g -O2 -DNDEBUG -g -fwrapv -O3 -Wall -Wstrict-prototypes -I/opt/homebrew/include -I/opt/homebrew/opt/openssl/include -I/opt/homebrew/opt/sqlite/include -I/opt/homebrew/Cellar/python27/2.7.10_2/Frameworks/Python.framework/Versions/2.7/include/python2.7 -c dulwich/_objects.c -o build/temp.macosx-10.11-x86_64-2.7/dulwich/_objects.o
    clang -bundle -undefined dynamic_lookup build/temp.macosx-10.11-x86_64-2.7/dulwich/_objects.o -L/opt/homebrew/lib -L/opt/homebrew/opt/openssl/lib -L/opt/homebrew/opt/sqlite/lib -o build/lib.macosx-10.11-x86_64-2.7/dulwich/_objects.so
    building 'dulwich._pack' extension
    clang -fno-strict-aliasing -fno-common -dynamic -g -O2 -DNDEBUG -g -fwrapv -O3 -Wall -Wstrict-prototypes -I/opt/homebrew/include -I/opt/homebrew/opt/openssl/include -I/opt/homebrew/opt/sqlite/include -I/opt/homebrew/Cellar/python27/2.7.10_2/Frameworks/Python.framework/Versions/2.7/include/python2.7 -c dulwich/_pack.c -o build/temp.macosx-10.11-x86_64-2.7/dulwich/_pack.o
    clang -bundle -undefined dynamic_lookup build/temp.macosx-10.11-x86_64-2.7/dulwich/_pack.o -L/opt/homebrew/lib -L/opt/homebrew/opt/openssl/lib -L/opt/homebrew/opt/sqlite/lib -o build/lib.macosx-10.11-x86_64-2.7/dulwich/_pack.so
    building 'dulwich._diff_tree' extension
    clang -fno-strict-aliasing -fno-common -dynamic -g -O2 -DNDEBUG -g -fwrapv -O3 -Wall -Wstrict-prototypes -I/opt/homebrew/include -I/opt/homebrew/opt/openssl/include -I/opt/homebrew/opt/sqlite/include -I/opt/homebrew/Cellar/python27/2.7.10_2/Frameworks/Python.framework/Versions/2.7/include/python2.7 -c dulwich/_diff_tree.c -o build/temp.macosx-10.11-x86_64-2.7/dulwich/_diff_tree.o
    clang -bundle -undefined dynamic_lookup build/temp.macosx-10.11-x86_64-2.7/dulwich/_diff_tree.o -L/opt/homebrew/lib -L/opt/homebrew/opt/openssl/lib -L/opt/homebrew/opt/sqlite/lib -o build/lib.macosx-10.11-x86_64-2.7/dulwich/_diff_tree.so
    changing mode of build/scripts-2.7/dul-receive-pack from 644 to 755
    changing mode of build/scripts-2.7/dul-upload-pack from 644 to 755
      File "/opt/homebrew/lib/python2.7/site-packages/dulwich/server.py", line 150
        def get_refs(self) -> Dict[bytes, bytes]:
                           ^
    SyntaxError: invalid syntax

      File "/opt/homebrew/lib/python2.7/site-packages/dulwich/web.py", line 70
        def date_time_string(timestamp: Optional[float] = None) -> str:
                                      ^
    SyntaxError: invalid syntax

      File "/opt/homebrew/lib/python2.7/site-packages/dulwich/porcelain.py", line 880
        repo: Repo,
            ^
    SyntaxError: invalid syntax

      File "/opt/homebrew/lib/python2.7/site-packages/dulwich/objects.py", line 154
        def object_header(num_type: int, length: int) -> bytes:
                                  ^
    SyntaxError: invalid syntax

    changing mode of /opt/homebrew/bin/dul-receive-pack to 755
    changing mode of /opt/homebrew/bin/dul-upload-pack to 755
    Installing dulwich script to /opt/homebrew/bin
  Found existing installation: urllib3 1.22
    Uninstalling urllib3:
      Successfully uninstalled urllib3
Successfully installed dulwich urllib3 certifi
Cleaning up...
```
(Projects using dulwish subsequently fail with the same SyntaxError).

My setup:
- Python 2.7.10
- Darwin matthiasb-mbp 19.4.0 Darwin Kernel Version 19.4.0: Wed Mar  4 22:28:40 PST 2020; root:xnu-6153.101.6~15/RELEASE_X86_64 x86_64

Using `pip2.7 install install dulwich==0.19.15` to work around the issue is working fine for me.

**Comment by jelmer** [comment]
Ah, thanks for the reminder - I've updated the list of supported versions on the homepage. 

I don't think there is a way to prevent pip2.7 from attempting to install Dulwich.

I'll see if I can get setup.py to raise an exception when run on python2.7, so at least the error is clearer.

**Comment by mjpieters** [comment]
> I don't think there is a way to prevent pip2.7 from attempting to install Dulwich.

You can set `python_requires` in `setuptools`:

```python
    python_requires='>=3.5',
```

**Comment by mjpieters** [comment]
Ah, you [already set this](https://github.com/dulwich/dulwich/blob/ba79362c88dc1e9b633d4b2ef78238d2f1ae28e8/setup.py#L89). But, because there is no _universal wheel_ (not an option because dulwich has C extensions), there is no matching wheel available for Python 2.7, so pip downloads the tar.gz, etc.

A possible solution is to add a `pyproject.toml` file that configures the build system to always use setuptools:

```ini
[build-system]
requires = ["setuptools >= 40.6.0", "wheel"]
build-backend = "setuptools.build_meta"
```

This tells pip to get `setuptools` too when installing. The added advantage is that you can then remove all the conditional code testing if `setuptools` is installed.
